File: utils/aggregation.ts
import { ETFData } from '@/types';

export const parseYYYYMMDD = (n: number) => {
  const y = Math.floor(n / 10000);
  const m = Math.floor((n % 10000) / 100) - 1;
  const d = n % 100;
  return new Date(y, m, d);
};

export const fmtYYYYMMDD = (d: Date) => {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return Number(`${y}${m}${day}`);
};

export const daysBetween = (a: Date, b: Date) => Math.max(1, Math.round((+b - +a) / 86400000));

export const chooseBucketDays = (spanDays: number, maxPoints = 60) => {
  let bucketDays: number;
  if (spanDays <= 60) bucketDays = 1;
  else if (spanDays <= 180) bucketDays = 7;
  else if (spanDays <= 720) bucketDays = 30;
  else bucketDays = 90;
  const est = Math.ceil(spanDays / bucketDays);
  return est > maxPoints ? Math.ceil(spanDays / maxPoints) : bucketDays;
};

export const aggregateOnBuckets = (
  rows: ETFData[],
  globalStart: Date,
  bucketDays: number,
  bucketCount: number
): { data: number[]; upOrDown: 'up' | 'down' } => {
  const sorted = [...rows].sort((a, b) => a.calendar_id - b.calendar_id);
  const acc = Array.from({ length: bucketCount }, () => ({ sum: 0, cnt: 0 }));

  for (const r of sorted) {
    const d = parseYYYYMMDD(r.calendar_id);
    let idx = Math.floor(daysBetween(globalStart, d) / bucketDays);
    if (idx < 0) idx = 0;
    if (idx >= bucketCount) idx = bucketCount - 1;
    const price = parseFloat(r.close_price);
    const cell = acc[idx];
    cell.sum += price;
    cell.cnt += 1;
  }

  const series: number[] = [];
  let prev = sorted.length ? parseFloat(sorted[0].close_price) : 0;
  for (let i = 0; i < bucketCount; i++) {
    const cell = acc[i];
    if (cell.cnt > 0) {
      prev = cell.sum / cell.cnt;
      series.push(prev);
    } else {
      series.push(prev);
    }
  }

  const first = series[0] ?? 0;
  const last = series[series.length - 1] ?? first;
  const upOrDown: 'up' | 'down' = last >= first ? 'up' : 'down';
  return { data: series, upOrDown };
};

export const buildBucketLabels = (globalStart: Date, bucketDays: number, bucketCount: number) => {
  const buildLabel = (d: Date) => {
    const y = d.getFullYear();
    const m = String(d.getMonth() + 1).padStart(2, '0');
    const day = String(d.getDate()).padStart(2, '0');
    return `${y}-${m}-${day}`;
  };
  const labels: string[] = [];
  for (let i = 0; i < bucketCount; i++) {
    const dt = new Date(globalStart.getTime() + i * bucketDays * 86400000);
    labels.push(buildLabel(dt));
  }
  return labels;
};

export default {
  parseYYYYMMDD,
  fmtYYYYMMDD,
  daysBetween,
  chooseBucketDays,
  aggregateOnBuckets,
  buildBucketLabels,
};

File: etf_flutter_app/lib/utils/aggregation.dart
int fmtYYYYMMDD(DateTime d) {
  final y = d.year;
  final m = (d.month).toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return int.parse('${y}${m}${day}');
}

DateTime parseYYYYMMDD(int n) {
  final y = n ~/ 10000;
  final m = (n % 10000) ~/ 100;
  final d = n % 100;
  return DateTime(y, m, d);
}

int daysBetween(DateTime a, DateTime b) => (b.difference(a).inDays).abs().clamp(1, 999999);

int chooseBucketDays(int spanDays, [int maxPoints = 60]) {
  int bucketDays;
  if (spanDays <= 60) bucketDays = 1;
  else if (spanDays <= 180) bucketDays = 7;
  else if (spanDays <= 720) bucketDays = 30;
  else bucketDays = 90;
  final est = (spanDays / bucketDays).ceil();
  return est > maxPoints ? (spanDays / maxPoints).ceil() : bucketDays;
}

class _AccCell {
  double sum = 0;
  int cnt = 0;
}

Map<String, dynamic> aggregateOnBuckets(List rows, DateTime globalStart, int bucketDays, int bucketCount) {
  final sorted = List.of(rows);
  sorted.sort((a, b) => (a.calendarId as int).compareTo(b.calendarId as int));
  final acc = List.generate(bucketCount, (_) => _AccCell());

  for (final r in sorted) {
    final d = parseYYYYMMDD(r.calendarId as int);
    var idx = (daysBetween(globalStart, d) / bucketDays).floor();
    if (idx < 0) idx = 0;
    if (idx >= bucketCount) idx = bucketCount - 1;
    final price = double.tryParse(r.closePrice.toString()) ?? 0.0;
    final cell = acc[idx];
    cell.sum += price;
    cell.cnt += 1;
  }

  final List<double> series = [];
  double prev = sorted.isNotEmpty ? double.tryParse(sorted[0].closePrice.toString()) ?? 0.0 : 0.0;
  for (var i = 0; i < bucketCount; i++) {
    final cell = acc[i];
    if (cell.cnt > 0) {
      prev = cell.sum / cell.cnt;
      series.add(prev);
    } else {
      series.add(prev);
    }
  }

  final first = series.isNotEmpty ? series[0] : 0.0;
  final last = series.isNotEmpty ? series.last : first;
  final upOrDown = last >= first ? 'up' : 'down';
  return {'data': series, 'upOrDown': upOrDown};
}

List<String> buildBucketLabels(DateTime globalStart, int bucketDays, int bucketCount) {
  String buildLabel(DateTime d) => '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
  final labels = <String>[];
  for (var i = 0; i < bucketCount; i++) {
    final dt = globalStart.add(Duration(days: i * bucketDays));
    labels.add(buildLabel(dt));
  }
  return labels;
}
